### ISSUE TITLE
Broken client response after `delete` within `multi`

### ISSUE THREAD

**Issue by igor04** [petergoldstein/dalli#772]
If we have invocation of `delete` with `key` which is no longer existed within `multi` then we will have broken client:
 
```ruby
require 'dalli'

a = Dalli::Client.new('localhost')
a.set('a', 'av')
a.set('b', 'bv')
p a.get('a') # => 'av'
p a.get('b') # => 'bv

a.multi do
  a.delete('not_existed_key')
end

p a.get('a') # => nil
p a.get('b') # => 'av'
p a.get('b') # => 'bv'
```
From example we could see that response is shifted and  it happens because `deleteq` inside `multi` still has response `Not found` but such response is not expcted and is not read from socket on `delete` invocation, but will be read only in place of next invocation (ex: `get`) and will result into broken client responses

**Comment by petergoldstein** [comment]
I've added a test for this behavior here - https://github.com/petergoldstein/dalli/pull/799 - and confirmed that the reported behavior is still a problem.

**Comment by petergoldstein** [comment]
@mperham I wanted to get your thoughts on this, as I believe you wrote the original code for multi.

So as I understand it, the write commands (setq, addq, replaceq, deleteq) potentially invoked in a multi block have a conditional response - they respond only if there's an error.  Effectively that means that (race conditions aside), when we exit the block we should either:

1. Have no responses (which means everything processed ok)
2. Have responses in the buffer that should be cleared out when we exit the multi block, possibly returning an error to the caller (although frankly I'm not sure where we'd do that, given the client structure)

But the race conditions I note above complicate this dramatically, since essentially we don't know that we're "done" with any of these operations (setq, addq, replaceq, deleteq).  In theory we could do a non-blocking read on the socket when we exit the `yield` in the `multi` block, but that still leaves the race condition. 

So as far as I can tell, using a `multi` block as in the above example always leaves the underlying socket in a potentially inconsistent state, with no way to guarantee a clean state at next command invocation.

Your thoughts?

**Comment by mperham** [comment]
I believe tmm1 rewrote most of the multi code when he was optimizing GitHub years ago.

Honestly it's been 5+ years since I've used Dalli or looked at the multi code but I can speak in generalities. Typically you'd give an OperationID with each request and any responses will be tagged with the corresponding OperationID so you know which operations had no response and which had an error. In general though, multi should drain the socket at the end of the block.

**Comment by mperham** [comment]
I don't believe multi has async semantics. Once you finish the last operation write, you can read the socket and get all of the responses at once.

**Comment by petergoldstein** [comment]
This is addressed in version 3.0.5.  Closing.
